includes/amp-helper-functions.php

```diff
@@ -18,8 +18,11 @@ function amp_get_permalink( $post_id ) {
 }
 
 function post_supports_amp( $post ) {
-	// Because `add_rewrite_endpoint` doesn't let us target specific post_types :(
-	if ( ! post_type_supports( $post->post_type, AMP_QUERY_VAR ) ) {
+	// Listen to post types settings.
+	$is_enabled_via_setting = (bool) AMP_Settings_Post_Types::get_instance()->get_settings_value( $post->post_type );
```

Comment:
Question: Instead of sourcing the enabled state via the settings directly, why not just add a separate action hook  that then just calls `add_post_type_support()` for each post type enabled be admin settings. This would allow plugins to query for AMP post type support via just `post_type_supports()` calls, and it could allow a plugin to turn off post type support for AMP via just `remove_post_type_support()`. Then this `post_supports_amp()` function could be left unchanged, as the plugin would essentially be providing a UI for doing the same `add_post_type_support()` calls that existing themes/plugins are doing today.

Reply:
Good question, that is how I had it initially but here is the issue I faced. If `post_type_supports()` is set by a theme/plugin then the admin setting must be checked and disabled according to the AC. The `post_type_supports()` support is usually declared in a `init` action which runs before the admin checks so if we add `post_type_supports` based on the db value saved then the checkbox becomes disabled (unless we declare it in an action running later which doesn't feel right). So I added a flag to differentiate the `post_type_supports()` that was added by the setting and the one that would be declared by a theme/plugin to prevent this behaviour. The issue is that if one is saved in the db and then a plugin/theme add support for it, then it won't be following the intended behaviour.
I will take another stab at this right now and post and update.

Reply:
Alright, [this commit](https://github.com/Automattic/amp-wp/pull/811/commits/09d3685989b114173389895614f9927a51947cb2) declare the `post_type_supports()` based on the admin settings. To do so, it runs it in a `after_setup_theme` callback so that it can be overwritten by theme/plugins which would declare `add_post_type_support()` or `remove_post_type_support()` invoked through an `init` action. 

Regarding the admin option disabled state, since we can't flag a post type which is not enabled by setting and removed by plugin/theme we can't disable the checkbox but the `AMP_Settings_Post_Types::errors()` takes care of this scenario if a user try to enable a post type support but a theme/plugin declares `remove_post_type_support()`.

One scenario that could be seen as an issue is if a plugin doesn’t have `add_post_type_support()` in its current version and the AMP settings are saved. If in the newer version the plugin add `add_post_type_support()` then the checkbox in the admin wouldn’t be disabled since it is flagged that the post type support is managed by the setting is saved in the db. For that reason, I would be in favor of removing the "inconsistent" disabled state and add an error handling in case a plugin/theme calls `add_post_type_support()` and a user tries to disable it. That would be a complete bullet proof solution. Thoughts?

Reply:
> If in the newer version the plugin add add_post_type_support() then the checkbox in the admin wouldn’t be disabled since it is flagged that the post type support is managed by the setting is saved in the db

What if the plugin instead had a constant that defined the post types that the AMP plugin supports by default? This could then be used in the `amp_core_post_types_support()`  function, and then when deciding whether or not to disable the checkbox it can look at the constant rather than (or in addition to?) doing the `post_type_supports()` call?

Reply:
> One scenario that could be seen as an issue is if a plugin doesn’t have `add_post_type_support()` in its current version and the AMP settings are saved. If in the newer version the plugin add `add_post_type_support()` then the checkbox in the admin wouldn’t be disabled since it is flagged that the post type support is managed by the setting is saved in the db
Plugin wouldn't be able to change the constant unless we add a filter, hence my suggestion to add a filter via PM earlier this week. That said, this would take care of plugins which would want to declare `remove_post_type_support()` but have the checkbox disabled which is a pity if we are going to root of adding an extra flag.

If we are going to add an extra step, then I would vote for a filter like `amp_post_types_support_checkbox_disabled`. That way we can let the error handling taking care of overwrite restrictions on save if the plugin didn't add `amp_post_types_support_checkbox_disabled` and allow plugins to force disable the checkbox to enhance user experience. Alternatively, we could just not have the disabled state and let error handle all scenarios on save which will be functionality bullet proof and consistent. Thoughts?

Reply:
I'm having a hard time running through the scenarios in my head. Can we create a little sample plugin that registers some post types, enables support for some of them, and disables support for others?

Reply:
@westonruter following on our discussion at WCUS, I pushed a commit which removed the disabled state and handle all scenarios using an errors handling on save. This covers all type of scenarios in a consistent and bullet proof way.

While it was nice to have the disable state on load from a user experience perspective, I believe the fact that it was so inconsistent (no way around it) didn't justify to have it and actually turned into a bad user experience as no one could really figure out what was going on. Please take it for a spin and let me know what you think.